File: src/resources/reference-data.ts
import { ResourceTemplate } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { ToolContext } from "../tools/common.js";

export function registerReferenceResources(server: any, ctx: ToolContext): void {
  server.registerResource("reference_statuses", "projeqtor://reference/statuses", { title: "Statuses", description: "Available ProjeQtOr statuses", mimeType: "application/json" }, async (uri: URL) => ({
    contents: [{ uri: uri.href, mimeType: "application/json", text: JSON.stringify(await ctx.client.listAll("Status"), null, 2) }]
  }));

  server.registerResource("reference_types", "projeqtor://reference/types", { title: "Types", description: "Project, activity and ticket types", mimeType: "application/json" }, async (uri: URL) => {
    const data = {
      projectTypes: await ctx.client.listAll("ProjectType").catch(() => []),
      activityTypes: await ctx.client.listAll("ActivityType").catch(() => []),
      ticketTypes: await ctx.client.listAll("TicketType").catch(() => []),
      priorities: await ctx.client.listAll("Priority").catch(() => [])
    };
    return { contents: [{ uri: uri.href, mimeType: "application/json", text: JSON.stringify(data, null, 2) }] };
  });

  server.registerResource(
    "resource_availability",
    new ResourceTemplate("projeqtor://resources/{id}/availability", { list: undefined }),
    { title: "Resource availability", description: "Resource assignments, work and leave periods for availability analysis", mimeType: "application/json" },
    async (uri: URL, vars: any) => {
      const id = Array.isArray(vars.id) ? vars.id[0] : vars.id;
      const data = {
        resource: await ctx.client.getObject("Resource", id),
        assignments: await ctx.client.search("Assignment", [{ field: "idResource", operator: "=", value: id }]).catch(() => []),
        work: await ctx.client.search("Work", [{ field: "idResource", operator: "=", value: id }]).catch(() => []),
        leaves: await ctx.client.search("LeavePeriod", [{ field: "idResource", operator: "=", value: id }]).catch(() => [])
      };
      return { contents: [{ uri: uri.href, mimeType: "application/json", text: JSON.stringify(data, null, 2) }] };
    }
  );
}

File: src/projeqtor_mcp_server/resources/reference_data.py
from __future__ import annotations

import json
from typing import Any

from mcp.server.fastmcp import FastMCP

from projeqtor_mcp_server.client.projeqtor_api import ProjeQtOrApiClient, SearchCriterion
from projeqtor_mcp_server.tools.common import safe


def register_reference_resources(mcp: FastMCP, client: ProjeQtOrApiClient) -> None:
    """Register reference data resources."""

    @mcp.resource("projeqtor://reference/statuses")
    async def reference_statuses() -> str:
        """Liste des statuts ProjeQtOr."""
        return json.dumps(await safe(client.list_all("Status")), ensure_ascii=False, indent=2)

    @mcp.resource("projeqtor://reference/types")
    async def reference_types() -> str:
        """Types projet/activité/ticket et priorités."""
        data: dict[str, Any] = {
            "projectTypes": await safe(client.list_all("ProjectType")),
            "activityTypes": await safe(client.list_all("ActivityType")),
            "ticketTypes": await safe(client.list_all("TicketType")),
            "priorities": await safe(client.list_all("Priority")),
        }
        return json.dumps(data, ensure_ascii=False, indent=2)

    @mcp.resource("projeqtor://resources/{id}/availability")
    async def resource_availability(id: str) -> str:
        """Disponibilité d'une ressource: affectations, work logs, congés."""
        data: dict[str, Any] = {
            "resource": await safe(client.get_object("Resource", id)),
            "assignments": await safe(client.search("Assignment", [SearchCriterion("idResource", id)])),
            "work": await safe(client.search("Work", [SearchCriterion("idResource", id)])),
            "leaves": await safe(client.search("LeavePeriod", [SearchCriterion("idResource", id)])),
        }
        return json.dumps(data, ensure_ascii=False, indent=2)
